x/tokenfilter/ibc_middleware.go

```diff
@@ -0,0 +1,81 @@
+package tokenfilter
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+)
+
+var _ porttypes.Middleware = &tokenFilterMiddleware{}
+
+const ModuleName = "tokenfilter"
+
+// tokenFilterMiddleware directly inherits the IBCModule and ICS4Wrapper interfaces.
+// Only with OnRecvPacket, does it wrap the underlying implementation with additional
+// stateless logic for rejecting the inbound transfer of non-native tokens. This
+// middleware is unilateral and no handshake is required. If using this middleware
+// on an existing chain, tokens that have been routed through this chain will still
+// be allowed to unwrap.
+type tokenFilterMiddleware struct {
+	porttypes.IBCModule
+	porttypes.ICS4Wrapper
+}
```

Comment:
Kinda nice trick with the embedding to avoid code boilerplate. Was this the intention here?

I do like the idea of not having to have a `keeper` package, given its a stateless middleware. 
But I would've expected to see the transfer keeper composed with something like `app.TokenFilterKeeper`.

```diff
app.TransferKeeper = ibctransferkeeper.NewKeeper(
	appCodec, keys[ibctransfertypes.StoreKey], app.GetSubspace(ibctransfertypes.ModuleName),
-	app.IBCKeeper.ChannelKeeper, app.IBCKeeper.ChannelKeeper, &app.IBCKeeper.PortKeeper,
+       app.TokenFilterKeeper, app.IBCKeeper.ChannelKeeper, &app.IBCKeeper.PortKeeper,
	app.AccountKeeper, app.BankKeeper, scopedTransferKeeper,
)
```
At the moment its completely bypassing this tokenfilter middleware for `SendPacket`. With the above, the `TokenFilterKeeper` would just be a simple struct composed with the ibc channel keeper.

E.g. for packets being received the flow is: `ibc core -> token filter -> transfer` where token filter essentially inherits transfers callbacks but overrides `OnRecvPacket`. When transfer sends a packet, the current flow is: `transfer -> ibc core`. 

While its technically fine for this right now, and token filter doesn't care about outbound packets I think the configuration might be a little bit error prone, for example, the `ICS4Wrapper` in this struct is never actually used. It would definitely become apparent especially if you try to add more middlewares to the transfer application stack in future. 

Does this make sense?

Reply:
> So because I just reference the ChannelKeeper directly, any middleware below this would get bypassed?

Yes, because `ChannelKeeper` is passed directly as the `ICS4Wrapper` arg to `NewTransferKeeper` then `SendPacket` would bypass all middlewares in the application stack. 

> Also, I'm not sure if I really care about SendPacket in the case of the TokenFilter and I would like to avoid having a keeper if possible.

Yeah I understand wanting to avoid a keeper, and like you say, `SendPacket` is essentially going to be a passthrough anyways and not implement any logic. 
I guess its just a general concern that you should be aware of with the current way its wired up, if you decide to make any changes or add additional middlewares in future which will want to implement logic on `SendPacket`. 

I would probably advise for taking the approach where the token filter is invoked bi-directionally, even if its a no-op. 
i.e.

```
recv: ibc core -> token filter -> transfer
send: transfer -> token filter -> ibc core
```

This way you avoid any future errors made my devs who are unaware of the call flow. 

All that being said, I understand your reasoning for avoiding boilerplate and the need for a keeper package. As long as you're aware of the risks then feel free to go ahead with this.

Reply:
fwiw you could define an `ICS4Wrapper` struct within this package, and use that as the arg in `NewTransferKeeper`. You'd avoiding having to add a keeper package that way and still stick to convention and adhere to the correct flows

Reply:
Ok I've given what you suggested a go just for conventions sake. 

It's does strike me however a bit strange that `Middleware` is both `IBCModule` and `ICS4Wrapper`. This is kind of impossible to implement within the same struct else you'd end with circular dependencies. In this case transfer keeper needs to both import token filter and be imported by token filter.